File: src/NodeParser/TypeReferenceNodeParser.ts
import ts from "typescript";
import { Context, type NodeParser } from "../NodeParser.js";
import type { SubNodeParser } from "../SubNodeParser.js";
import { AnnotatedType } from "../Type/AnnotatedType.js";
import { AnyType } from "../Type/AnyType.js";
import { ArrayType } from "../Type/ArrayType.js";
import type { BaseType } from "../Type/BaseType.js";
import { StringType } from "../Type/StringType.js";
import { UnknownType } from "../Type/UnknownType.js";
import { UnhandledError } from "../Error/Errors.js";
import { symbolAtNode } from "../Utils/symbolAtNode.js";

const invalidTypes: Record<number, boolean> = {
    [ts.SyntaxKind.ModuleDeclaration]: true,
    [ts.SyntaxKind.VariableDeclaration]: true,
};

export class TypeReferenceNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.TypeReferenceNode): boolean {
        return node.kind === ts.SyntaxKind.TypeReference;
    }

    public createType(node: ts.TypeReferenceNode, context: Context): BaseType {
        const typeSymbol =
            this.typeChecker.getSymbolAtLocation(node.typeName) ??
            // When the node doesn't have a valid source file, its position is -1, so we can't
            // search for a symbol based on its location. In that case, the ts.factory defines a symbol
            // property on the node itself.
            symbolAtNode(node.typeName)!;

        if (typeSymbol.flags & ts.SymbolFlags.Alias) {
            const aliasedSymbol = this.typeChecker.getAliasedSymbol(typeSymbol);

            const declaration = aliasedSymbol.declarations?.filter((n: ts.Declaration) => !invalidTypes[n.kind])[0];

            if (!declaration) {
                // fallback for bun.sh
                return new AnyType();
            }

            return this.createTypeFromDeclaration(declaration, node, context);
        }

        if (typeSymbol.flags & ts.SymbolFlags.TypeParameter) {
            return context.getArgument(typeSymbol.name) ?? new UnknownType(true);
        }

        // Wraps promise type to avoid resolving to a empty Object type.
        if (typeSymbol.name === "Promise" || typeSymbol.name === "PromiseLike") {
            // Promise without type resolves to Promise<any>
            if (!node.typeArguments || node.typeArguments.length === 0) {
                return new AnyType();
            }

            return this.childNodeParser.createType(node.typeArguments[0], context);
        }

        if (typeSymbol.name === "Array" || typeSymbol.name === "ReadonlyArray") {
            const type = this.createSubContext(node, context).getArguments()[0];

            return type === undefined ? new AnyType() : new ArrayType(type);
        }

        if (typeSymbol.name === "Date") {
            return new AnnotatedType(new StringType(), { format: "date-time" }, false);
        }

        if (typeSymbol.name === "RegExp") {
            return new AnnotatedType(new StringType(), { format: "regex" }, false);
        }

        if (typeSymbol.name === "URL") {
            return new AnnotatedType(new StringType(), { format: "uri" }, false);
        }

        return this.createTypeFromDeclaration(
            typeSymbol.declarations!.filter((n: ts.Declaration) => !invalidTypes[n.kind])[0],
            node,
            context,
        );
    }

    /**
     * Resolves a referenced declaration through the child parser, falling back to the
     * type checker's already-resolved type when structural re-parsing crashes.
     *
     * Re-parsing the AST can drive the parser into type-level machinery of third-party
     * libraries (e.g. `zod`'s `z.infer<typeof schema>` conditional types, whose return
     * type is built from deep generics) that cannot be statically re-derived. When that
     * happens an unexpected (non-controlled) error is thrown; in that case we delegate to
     * the type checker, which has already resolved such references to a concrete type.
     */
    protected createTypeFromDeclaration(
        declaration: ts.Declaration,
        node: ts.TypeReferenceNode,
        context: Context,
    ): BaseType {
        try {
            return this.childNodeParser.createType(declaration, this.createSubContext(node, context));
        } catch (error) {
            if (error instanceof UnhandledError) {
                const resolved = this.createTypeFromChecker(node, context);
                if (resolved) {
                    return resolved;
                }
            }
            throw error;
        }
    }

    /**
     * Builds a base type from the type checker's fully-resolved type for the given
     * reference node. Returns undefined when the checker cannot offer a trustworthy
     * concrete type (e.g. it collapses to `any`/`unknown`), so callers can rethrow the
     * original error instead of silently substituting a meaningless schema.
     */
    protected createTypeFromChecker(node: ts.TypeReferenceNode, context: Context): BaseType | undefined {
        let resolvedType: ts.Type;
        try {
            resolvedType = this.typeChecker.getTypeFromTypeNode(node);
        } catch {
            return undefined;
        }

        const typeNode = this.typeChecker.typeToTypeNode(resolvedType, node, ts.NodeBuilderFlags.IgnoreErrors);

        if (!typeNode || typeNode.kind === ts.SyntaxKind.AnyKeyword || typeNode.kind === ts.SyntaxKind.UnknownKeyword) {
            return undefined;
        }
        try {
            const result = this.childNodeParser.createType(typeNode, context);
            if (result instanceof AnyType || result instanceof UnknownType) {
                return undefined;
            }
            return result;
        } catch {
            return undefined;
        }
    }

    protected createSubContext(node: ts.TypeReferenceNode, parentContext: Context): Context {
        const subContext = new Context(node);

        if (node.typeArguments?.length) {
            for (const typeArg of node.typeArguments) {
                subContext.pushArgument(this.childNodeParser.createType(typeArg, parentContext));
            }
        }

        return subContext;
    }
}

File: internal/parser/type_reference_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/tsutils"
	"github.com/vega/ts-json-schema-generator/internal/types"
)

// invalidTypes lists declaration kinds that are skipped when picking a
// symbol's declaration (src/NodeParser/TypeReferenceNodeParser.ts).
var invalidTypes = map[ast.Kind]bool{
	ast.KindModuleDeclaration:   true,
	ast.KindVariableDeclaration: true,
}

func firstValidDeclaration(declarations []*ast.Node) *ast.Node {
	for _, declaration := range declarations {
		if !invalidTypes[declaration.Kind] {
			return declaration
		}
	}
	return nil
}

// TypeReferenceNodeParser resolves type references
// (src/NodeParser/TypeReferenceNodeParser.ts).
type TypeReferenceNodeParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewTypeReferenceNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser) *TypeReferenceNodeParser {
	return &TypeReferenceNodeParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *TypeReferenceNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindTypeReference
}

func (p *TypeReferenceNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	typeName := node.AsTypeReferenceNode().TypeName

	typeSymbol := p.typeChecker.GetSymbolAtLocation(typeName)
	if typeSymbol == nil {
		// When the node doesn't have a valid source file, its position is -1,
		// so we can't search for a symbol based on its location. In that case,
		// the factory defines a symbol property on the node itself.
		typeSymbol = tsutils.SymbolAtNode(typeName)
	}

	if typeSymbol.Flags&ast.SymbolFlagsAlias != 0 {
		aliasedSymbol := p.typeChecker.GetAliasedSymbol(typeSymbol)

		declaration := firstValidDeclaration(aliasedSymbol.Declarations)
		if declaration == nil {
			// fallback for bun.sh
			return &types.AnyType{}
		}

		return p.childNodeParser.CreateType(declaration, p.createSubContext(node, ctx), nil)
	}

	if typeSymbol.Flags&ast.SymbolFlagsTypeParameter != 0 {
		if argument := ctx.GetArgument(typeSymbol.Name); argument != nil {
			return argument
		}
		return &types.UnknownType{ErroredSource: true}
	}

	// Wraps promise type to avoid resolving to an empty Object type.
	if typeSymbol.Name == "Promise" || typeSymbol.Name == "PromiseLike" {
		typeArguments := node.TypeArguments()

		// Promise without type resolves to Promise<any>.
		if len(typeArguments) == 0 {
			return &types.AnyType{}
		}

		return p.childNodeParser.CreateType(typeArguments[0], ctx, nil)
	}

	if typeSymbol.Name == "Array" || typeSymbol.Name == "ReadonlyArray" {
		arguments := p.createSubContext(node, ctx).Arguments()
		if len(arguments) == 0 || arguments[0] == nil {
			return &types.AnyType{}
		}
		return &types.ArrayType{Item: arguments[0]}
	}

	switch typeSymbol.Name {
	case "Date":
		return &types.AnnotatedType{Type: &types.StringType{}, Annotations: types.Annotations{"format": "date-time"}, Nullable: false}
	case "RegExp":
		return &types.AnnotatedType{Type: &types.StringType{}, Annotations: types.Annotations{"format": "regex"}, Nullable: false}
	case "URL":
		return &types.AnnotatedType{Type: &types.StringType{}, Annotations: types.Annotations{"format": "uri"}, Nullable: false}
	}

	return p.childNodeParser.CreateType(firstValidDeclaration(typeSymbol.Declarations), p.createSubContext(node, ctx), nil)
}

func (p *TypeReferenceNodeParser) createSubContext(node *ast.Node, parentContext *Context) *Context {
	subContext := NewContext(node)
	for _, typeArg := range node.TypeArguments() {
		subContext.PushArgument(p.childNodeParser.CreateType(typeArg, parentContext, nil))
	}
	return subContext
}
